internal/controller/promotion/argocd.go

```diff
@@ -301,22 +312,46 @@ func (a *argoCDMechanism) mustPerformUpdate(
 	}
 
 	// Check if the desired revision was applied.
-	if desiredRevision, err := libargocd.GetDesiredRevision(
+	desiredRevisions, err := libargocd.GetDesiredRevisions(
 		ctx,
 		a.kargoClient,
 		stage,
 		update,
 		app,
 		newFreight,
-	); err != nil {
+	)
+	appLogger.Debug("Operation completed, checking applied revision.", "desiredRevisions", desiredRevisions)
+
+	if err != nil {
 		return "", true, fmt.Errorf("error determining desired revision: %w", err)
-	} else if desiredRevision != "" && status.SyncResult.Revision != desiredRevision {
-		// The operation did not result in the desired revision being applied.
-		// We should attempt to retry the operation.
+	}
+
+	// Check multi-source app for matching revisions.
+	if len(status.SyncResult.Revisions) > 0 {
+		matchingRevisions := libargocd.GetIntersection(status.SyncResult.Revisions, desiredRevisions)
+		if len(matchingRevisions) == len(desiredRevisions) {
```

Comment:
You're never doing anything with `matchingRevisions` other than checking its length. It is possible the question you are looking to have answered (thought it _may involve_ finding the intersection of observed and desired revisions) is a yes/no of "are all source synced to the desired revision?"

Reply:
We're also logging this in eg `health.go` line 303 where it matters, I left a more detailed breakdown in reply to your comment under the function, I'm closing this one to avoid flip-flopping between sections.